### PR TITLE
Register sdpa, transformers reference implementation as golden functions

### PR DESCRIPTION
**Pull request by ctr-martemovTT**
### PR Category

Part of the implementation for https://github.com/tenstorrent/tt-metal/issues/28214 -- move existing reference function implementations for the SDPA operations in the shared modules, and provide the tests for the golden functions. 

Follow-up for https://github.com/tenstorrent/tt-metal/pull/56748

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:martemov/comparison-mode-fixes-5)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:martemov/comparison-mode-fixes-5)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:martemov/comparison-mode-fixes-5)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:martemov/comparison-mode-fixes-5)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-tests.yaml?query=branch:martemov/comparison-mode-fixes-5)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:martemov/comparison-mode-fixes-5)

### CODE REVIEW

**Comment by yieldthought** [comment]
Codex review:

1. **[P2] Preserve the query batch when decoding a shared cache** — [ttnn/ttnn/operations/transformer_golden.py:433](https://github.com/tenstorrent/tt-metal/blob/0929f5d02048b1eec534ee60609a9856e1bbe1d4/ttnn/ttnn/operations/transformer_golden.py#L433)

   With `share_cache=True`, K/V have batch 1 while Q is `[1, B, H, D]`. This check leaves Q untransposed when `B > 1`, so `_decode_attention` treats the batch as heads and processes only `cur_pos[0]`. A host case with Q `[1,2,32,64]`, K/V `[1,1,64,64]`, and positions `[63,31]` returns the expected shape but differs from independent per-user SDPA by a maximum absolute error of `0.79385`. Suggestion: normalize the public decode query layout independently of the cache batch and retain each user's position/mask.

2. **[P2] Normalize mixed sparse-attention dtypes before the contractions** — [ttnn/ttnn/operations/transformer_golden.py:788](https://github.com/tenstorrent/tt-metal/blob/0929f5d02048b1eec534ee60609a9856e1bbe1d4/ttnn/ttnn/operations/transformer_golden.py#L788)

   `sparse_sdpa` supports BF16 Q with raw FP8 KV (and the reverse), but comparison preprocessing converts FP8 to `torch.float32` while BF16 stays BF16. Passing those inputs to this newly registered reference raises `expected m1 and m2 to have the same dtype`; the final value contraction also assumes matching dtypes. I reproduced this with BF16 Q and FP8-quantized KV converted to float32. Consequently comparison is skipped, or raises in strict mode, for a supported operation configuration. Suggestion: use a common computation dtype for both contractions and preserve the intended output dtype.

3. **[P2] Decode the packed sparse KV format before using the dense reference** — [ttnn/ttnn/operations/transformer_golden.py:814](https://github.com/tenstorrent/tt-metal/blob/0929f5d02048b1eec534ee60609a9856e1bbe1d4/ttnn/ttnn/operations/transformer_golden.py#L814)

   The required `kv_format` argument is swallowed by `**_`. For `SparseKVFormat.SCALED_FP8`, the cache contains 656-byte rows (`512 FP8 + 4 FP32 scales + 64 BF16 RoPE`), while Q's key dimension is 576. Passing this storage directly to `sparse_mla` fails at the expansion to `[B,T,576]` because KV is `[B,T,656]`; host conversion preserves that width. This affects the existing scaled-FP8 sparse-SDPA callers. Suggestion: add format-aware preprocessing that reconstructs the scaled latent and BF16 RoPE before calling the reference, preserving the packed bytes through conversion.

4. **[P2] Give the new ring-MLA tests matching Q/K dimensions** — [tests/ttnn/unit_tests/operations/sdpa/test_sdpa_golden_functions.py:456](https://github.com/tenstorrent/tt-metal/blob/0929f5d02048b1eec534ee60609a9856e1bbe1d4/tests/ttnn/unit_tests/operations/sdpa/test_sdpa_golden_functions.py#L456) and [tests/ttnn/unit_tests/operations/sdpa/test_sdpa_golden_functions.py:475](https://github.com/tenstorrent/tt-metal/blob/0929f5d02048b1eec534ee60609a9856e1bbe1d4/tests/ttnn/unit_tests/operations/sdpa/test_sdpa_golden_functions.py#L475)

   Both tests construct Q with width 4 and KV with width 6. `ring_mla_golden` uses the entire KV row as K, so both fail inside PyTorch SDPA before reaching the cache-selection assertions (`[8,4]` versus `[8,6]`, and `[2,4]` versus `[2,6]`). Suggestion: change the query widths to 6. I verified both test bodies pass with those two edits applied only in memory; the inline suggestions contain the exact changes.

Host verification at `0929f5d02048b1eec534ee60609a9856e1bbe1d4`: 23 of 25 new test bodies passed, with the two failures above. Native TTNN registration was represented by a minimal stand-in for these host runs; this was not a device or end-to-end comparison-mode test.

File: tests/ttnn/unit_tests/operations/sdpa/test_sdpa_golden_functions.py [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+        gathered_value.repeat_interleave(2, dim=1),
+        attn_mask=logical_mask[..., 4:6],
+        scale=1.0,
+    ).permute(2, 0, 1, 3)
+    torch.testing.assert_close(actual, expected)
+
+
+def test_ring_mla_golden_selects_cache_slot_when_batch_sizes_match():
+    torch.manual_seed(3)
+    query = torch.randn(4, 2, 3, 4)
```
Comment (yieldthought) [inline tests/ttnn/unit_tests/operations/sdpa/test_sdpa_golden_functions.py]: Codex: Match Q's width to the six-wide K row so this test reaches its cache-selection assertions. The test body passes with this change applied in memory.

```suggestion
    query = torch.randn(4, 2, 3, 6)
```

File: tests/ttnn/unit_tests/operations/sdpa/test_sdpa_golden_functions.py [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+        key.repeat_interleave(2, dim=1),
+        value.repeat_interleave(2, dim=1),
+        is_causal=True,
+    )
+    torch.testing.assert_close(output, expected)
+
+
+def test_ring_mla_golden_uses_runtime_slot_and_logical_prefix():
+    torch.manual_seed(9)
+    query = torch.randn(1, 2, 2, 4)
```
Comment (yieldthought) [inline tests/ttnn/unit_tests/operations/sdpa/test_sdpa_golden_functions.py]: Codex: Match Q's width to the six-wide K row so this test reaches its cache-selection assertions. The test body passes with this change applied in memory.

```suggestion
    query = torch.randn(1, 2, 2, 6)
```

File: ttnn/ttnn/operations/transformer_golden.py [hunk 3]
```diff
@@ -0,0 +1,1695 @@
...
+    if positions is None or (not isinstance(positions, torch.Tensor) and len(positions) == 0):
+        positions = [input_tensor_k.shape[-2] - 1] * batch
+    if isinstance(positions, torch.Tensor):
+        positions = positions.reshape(-1).tolist()
+
+    outputs = []
+    for batch_index in range(batch):
+        position = int(positions[batch_index % len(positions)])
+        if position < 0:
+            outputs.append(torch.zeros_like(query[batch_index : batch_index + 1]))
```
Comment (yieldthought) [inline ttnn/ttnn/operations/transformer_golden.py:482]: Codex: Suggestion: size the skipped row using V's width so MLA batches can combine active 512-wide outputs with inactive users. Verified in memory for paged and unpaged MLA with active, mixed, and inactive batches.

```suggestion
            outputs.append(query.new_zeros((1, num_heads, 1, input_tensor_v.shape[-1])))
```

**Comment by yieldthought** [comment]
Codex review:

1. **[P2] Match the ring statistics output contract** — [ttnn/ttnn/operations/transformer_golden.py:1232](https://github.com/tenstorrent/tt-metal/blob/2ed539e908211c56b5807e7dcfce4e38cefeb517/ttnn/ttnn/operations/transformer_golden.py#L1232) and [ttnn/ttnn/operations/transformer_golden.py:1306](https://github.com/tenstorrent/tt-metal/blob/2ed539e908211c56b5807e7dcfce4e38cefeb517/ttnn/ttnn/operations/transformer_golden.py#L1306)

   Both ring-joint variants return a statistics buffer with twice the padded query-plus-joint length. The golden instead returns one LSE per query: a 32-token query without joint tokens produces `[B,H,32,1]` here versus `[B,H,64,1]` from `compute_output_specs`. `ring_mla_golden` doubles the LSE to match the size, but the streaming path uses this buffer for intermediate max/sum scratch, not duplicated final LSE. Comparison checks this output too, so valid attention calls fail comparison. Suggestion: give scratch outputs an explicit comparison-skip policy, or model their actual contract, and test through the comparison wrapper.

2. **[P2] Apply the block-cyclic KV index mapping** — [ttnn/ttnn/operations/transformer_golden.py:847](https://github.com/tenstorrent/tt-metal/blob/2ed539e908211c56b5807e7dcfce4e38cefeb517/ttnn/ttnn/operations/transformer_golden.py#L847)

   `block_cyclic_sp_axis`, `block_cyclic_chunk_local`, and `block_cyclic_cache_tp_sharded` disappear into `**_`, so natural token indices are applied directly to physical KV rows. The existing multi-device sparse-SDPA test uses a replicated cache whose rows remain in block-cyclic order after host conversion. With SP=2, T=256, and chunk_local=32, physical row 32 holds natural token 64. A host probe selecting only natural token 32 returns 64 instead of 32 when values encode token positions. Suggestion: preserve the mesh layout metadata during preprocessing and remap indices or restore natural cache order before evaluating attention.

3. **[P2] Preserve packed FP8 bytes for global comparison as well** — [ttnn/ttnn/operations/transformer.py:231](https://github.com/tenstorrent/tt-metal/blob/2ed539e908211c56b5807e7dcfce4e38cefeb517/ttnn/ttnn/operations/transformer.py#L231)

   The custom preprocessor fixes local comparison only. With `report_path` set, `preprocess_global_golden_function_inputs` independently processes the original arguments and value-converts an uncached FP8 KV tensor to float32; the subsequent metadata merge does not replace tensor arguments. The new unpacker then raises `Scaled-FP8 sparse KV must keep its packed bytes`. Executing these preprocessing functions with transport stand-ins reproduces a successful local golden and a failing global golden for the same packed input. Suggestion: carry the format-aware byte preservation through global preprocessing and add a wrapper-level regression.

4. **[P2] Use the value width for inactive MLA decode rows** — [ttnn/ttnn/operations/transformer_golden.py:482](https://github.com/tenstorrent/tt-metal/blob/2ed539e908211c56b5807e7dcfce4e38cefeb517/ttnn/ttnn/operations/transformer_golden.py#L482)

   The supported `cur_pos_tensor=-1` skip path allocates zeros with Q's width. MLA commonly has Q/K width 576 and V width 512, so a mixed active/inactive batch fails at `torch.cat`: `Expected size 512 but got size 576`. I reproduced this for both paged and unpaged MLA decode with positions `[63,-1]`. Suggestion: allocate skipped rows with `input_tensor_v.shape[-1]`. The inline suggestion passes eight host controls covering both APIs and active, mixed, and inactive batches.

Validation at `2ed539e908211c56b5807e7dcfce4e38cefeb517`: all 30 added host test cases passed under PyTorch 2.11.0 with native TTNN registration/transport represented by stand-ins. The findings above come from additional host probes and tracing the native API/kernel contracts; no device tests were run.

**Review by bbradelTT — APPROVED** [review]
Looks ok overall.

Please comment on or address the following AI comments:

1. Ring-joint “LSE” is not the device stats tensor.
Device compute_output_specs documents a 2× padded Q+joint buffer: first half running max, second half running sum (scratch, not final LSE). The golden returns logsumexp with one row per query. Comparison mode will fail on that third output even when attention is correct. ring_mla_golden concatenates [lse, lse] to fake the size; ring-joint does not. Skip/mask that buffer, or match the scratch contract.

2. Inactive MLA decode uses Q width for zeros.
cur_pos < 0 does torch.zeros_like(query_row). MLA often has Q/K 576 and V 512, so a mixed active/inactive batch fails at torch.cat. Same bug in paged circular decode. Size skipped rows from input_tensor_v.shape[-1] (inline suggestion on the PR).

3. Sparse block_cyclic_* is ignored.
Those kwargs go into **_. Device indices are natural token ids into a physically block-cyclic cache. Host conversion keeps physical order, so comparison on the existing multi-device sparse tests will attend the wrong rows.

4. Packed FP8 is only preserved on the local golden path.
_preprocess_sparse_sdpa_golden_inputs unpacks after preserve_fp8_bytes=True. Global comparison (report_path) still runs preprocess_global_golden_function_inputs, which value-converts FP8 to float32; the unpacker then raises. Local comparison can pass while global comparison fails on the same call.

5. Either treat the ring stats tensor as skip/scratch or model max/sum explicitly.

**Review by arginugaTT — APPROVED** [review]

**Review by cglagovichTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/transformers/test_chunk_gated_delta_rule.py [hunk 4]
```diff
@@ -36,6 +36,85 @@
 PCC_STATE = 0.99999
 
 
+def test_gated_delta_rule_ops_have_registered_golden_functions():
+    assert callable(ttnn.get_golden_function(ttnn.transformer.chunk_gated_delta_rule))
+    assert callable(ttnn.get_golden_function(ttnn.transformer.gated_delta_attn_seq))
+
+
```
Comment (iwroszTT) [inline tests/ttnn/unit_tests/operations/transformers/test_chunk_gated_delta_rule.py:43]: Could you please update the shapes as follows (those shapes increase code coverage of the golden operation), change the formula for `expected` and add the additional test case:
```

B, T, HK, HV, K, V, CS = 2, 4, 2, 6, 2, 3, 2

def test_head_major_layout():
    token_major, state = golden(q, k, v, g, beta, chunk_size=CS, output_final_state=True)
    head_major, hm_state = golden(q, k, v, g, beta, chunk_size=CS, output_final_state=True,
                                  output_head_major=True)
    assert token_major.shape == (B, T, HV, V) and head_major.shape == (B * HV, T, V)
    # built independently of the implementation's permute/reshape:
    expected = torch.stack([token_major[b, :, h] for b in range(B) for h in range(HV)])
    torch.testing.assert_close(head_major, expected)
    torch.testing.assert_close(hm_state, state)

def test_gqa_expansion_is_repeat_interleave():
    got, got_state = golden(q, k, v, g, beta, chunk_size=CS, output_final_state=True)
    qe, ke = q.repeat_interleave(HV // HK, dim=2), k.repeat_interleave(HV // HK, dim=2)
    want, want_state = golden(qe, ke, v, g, beta, chunk_size=CS, output_final_state=True)
    torch.testing.assert_close(got, want)
    torch.testing.assert_close(got_state, want_state)
```

**Review by kpaigwar — APPROVED** [review]

File: ttnn/ttnn/decorators.py [hunk 5]
```diff
@@ -518,7 +519,7 @@ def _convert_ttnn_to_torch_for_comparison(tensor, **kwargs):
     return ttnn.to_torch(tensor, **kwargs)
```
Comment (vtsilytskyiTT) [inline ttnn/ttnn/decorators.py:519]: You can move whole body of this function inside `def convert(value, **kwargs):`. `_convert_ttnn_to_torch_for_comparison` isn't used anywhere else in this file as I can see.

**Review by vtsilytskyiTT — APPROVED** [review]
Core TTNN part approved. One small comment below.

**Review by iwroszTT — APPROVED** [review]
